### PR TITLE
feat(gatsby-plugin-sharp): create image sizes helper

### PR DESCRIPTION
**Pull request by gillkyle**
## Description

**TL:DR - utility that takes a bunch of info about an image and turns it into an array of numbers representing sizes we want to transform the image to**

To decouple some of the behavior in `gatsby-plugin-sharp` and make the image resizing easier to test and tweak, this creates a utility that accepts a variety of arguments about an image, specifically:

- `file`: some info like the path the image is stored at
- `imgDimensions`: specific width/height of the file on disk that represents the maximum sizes that images can be transformed to
- `width`/`height`/`maxWidth`/`maxHeight`: the desired dimensions of the image being transformed
- `layout`: can be one of three types (`fixed` | `fluid` | `constrained`) and helps decide which branch of logic to follow to create sizes to transform to.
- `outputPixelDensities`: array of desired multipliers to size up the image with for display on higher pixel density screens (**should the defaults here be `[1, 2, 3]`?**) 
- `srcSetBreakpoints`: array of pixel widths to additionally include in the output of sizes

I wrote up a variety of test cases to start trying to cover possibilities of how this could be used. I'm sure I've missed plenty however. 

## Related Issues

Related to work on this branch: https://github.com/gatsbyjs/gatsby/tree/feat/image-resolvers (hence the PR to merge into it haha)

### CODE REVIEW

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 1]
```diff
@@ -88,33 +88,48 @@ export function rgbToHex(red, green, blue) {
...
     if (!width) {
-      width = height * file.aspectRatio
+      width = height * aspectRatio
     }
 
     sizes = densities
+      .filter(size => size >= 1) // remove smaller densities because fixed images don't need them
       .map(density => density * width)
-      .filter(size => size <= width)
+      .filter(size => size <= imgDimensions.width)
```
Comment (laurieontech) [inline packages/gatsby-plugin-sharp/src/utils.js:130]: I know this is less than or equal but it looks like a reverse arrow and almost broke my brain haha

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 2]
```diff
@@ -88,33 +88,48 @@ export function rgbToHex(red, green, blue) {
...
   width,
   maxWidth,
   height,
   maxHeight,
   layout,
-  outputPixelDensities,
+  outputPixelDensities = [0.25, 0.5, 1, 2, 3],
   srcSetBreakpoints,
 }) => {
+  if (maxWidth === 0 || width === 0 || maxHeight === 0 || height === 0) {
```
Comment (ascorbic) [inline packages/gatsby-plugin-sharp/src/utils.js:103]: Check that they're positive too

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 3]
```diff
@@ -124,19 +139,22 @@ export const calculateImageSizes = ({
...
       }px!
                      If possible, replace the current image with a larger one.
                      `)
     }
-  } else {
+  } else if (layout === `fluid` || layout === `constrained`) {
     // FLUID
     // if no maxWidth is passed, we need to resize the image based on the passed maxHeight
     const fixedDimension = maxWidth === undefined ? `maxHeight` : `maxWidth`
     maxWidth = maxWidth
```
Comment (ascorbic) [inline packages/gatsby-plugin-sharp/src/utils.js:151]: This will fail if neither maxHeight or maxWidth is set, as it tries to multiply maxHeight by the aspect ratio. If neither is set, you should set the *width* to the default size and calculate the height from that

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 4]
```diff
@@ -146,29 +164,30 @@ export const calculateImageSizes = ({
...
+    if (srcSetBreakpoints) {
+      srcSetBreakpoints.forEach(breakpoint => sizes.push(breakpoint))
     }
+    sizes = sizes.filter(size => size <= imgDimensions.width)
+
+    // ensure that the size passed in is included in the final output
+    if (!sizes.includes(maxWidth)) {
+      sizes.push(maxWidth)
+    }
+    sizes = sizes.sort((a, b) => a - b)
```
Comment (ascorbic) [inline packages/gatsby-plugin-sharp/src/utils.js:186]: This is the default comparator, so you can just do `sizes.sort()`

File: packages/gatsby-plugin-sharp/src/__tests__/utils.js [hunk 5]
```diff
@@ -39,3 +40,190 @@ describe(`createGatsbyProgressOrFallbackToExternalProgressBar`, () => {
...
+      file,
+      imgDimensions,
+    }
+
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toEqual(expect.arrayContaining([50, 70, 150, 250]))
+    expect(sizes).toEqual(expect.not.arrayContaining([1250, 1500]))
+  })
+
+  it(`should also include sizes from srcSetBreakpoints when outputPixelDensities are also passed in`, () => {
```
Comment (ascorbic) [inline packages/gatsby-plugin-sharp/src/__tests__/utils.js:212]: I think if we're passing in srcsetBreakpoints we should _only_ use those

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 6]
```diff
@@ -146,29 +164,30 @@ export const calculateImageSizes = ({
...
-      sizes = sizes.sort()
-    } else {
-      let sizes = srcSetBreakpoints.filter(size => size <= maxWidth)
-      if (!sizes.includes(maxWidth)) {
-        sizes.push(maxWidth)
-      }
-      sizes = sizes.sort()
+    sizes = densities.map(density => density * fixedValue)
+    // add breakpoint set to sizes if they are specified
+    if (srcSetBreakpoints) {
```
Comment (ascorbic) [inline packages/gatsby-plugin-sharp/src/utils.js:177]: srcSetBreakpoints should override everything else

Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/utils.js:177]: Okay, in that case, do you think we should warn that any provided outputPixelDensities would have no effect if both are passed in? `srcSetBreakpoints` and `outputPixelDensities` approach the same kind of thing (custom specified sizes) from different angles. It seems a little confusing how they'd work together in that case.

The current resizing function adds the breakpoints to the array, but I can make it use them exclusively.

File: packages/gatsby-plugin-sharp/src/__tests__/utils.js (hunk 5 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/__tests__/utils.js:212]: (Related to https://github.com/gatsbyjs/gatsby/pull/27554#discussion_r507853223)

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 6 above)
Comment (ascorbic) [reply packages/gatsby-plugin-sharp/src/utils.js:177]: Yeah, I think it should be one or the other. A warning is a good idea. What do you reckon, @laurieontech ?

Comment (laurieontech) [reply packages/gatsby-plugin-sharp/src/utils.js:177]: Ya, people might miss it, but always good to let them know when their code is ignored

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 2 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/utils.js]: Just added some better checking on this in https://github.com/gatsbyjs/gatsby/pull/27554/commits/f90f4bfb5cb0377751694384234d2ed19c6477e5

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 6 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/utils.js:185]: Okay, I'll add a warning then, thanks! My suspicion is _most_ people won't bother to use either one 😅 but if they need to, it should be clear.

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 4 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/utils.js]: So turns out this has this fun behavior of sorting by character which is almost right 😅 :

```
[0, 10, 190, 120, 500, 5, 1000000].sort()
> [0, 10, 1000000, 120, 190, 5, 500]
```

vs.

```
[0, 10, 190, 120, 500, 5, 1000000].sort((a, b) => a - b)
> [0, 5, 10, 120, 190, 500, 1000000]
```

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 3 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/utils.js]: I _think_ I've handled this based on what I understand. It's not the most functional logic like this writing over values over and over but seems to get the job done: https://github.com/gatsbyjs/gatsby/pull/27554/commits/c399322dfffdac9757897956dbc4e865b70aab83#diff-36db1668fae19fc6f3b88538a610293a66071255a4633e383b76b9b7e79b97edR168-R181

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 6 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/utils.js:201]: Added a warning [in a new commit](https://github.com/gatsbyjs/gatsby/pull/27554/commits/e4215704e77b6b8190dc7fd423d18815ab68b52e#diff-36db1668fae19fc6f3b88538a610293a66071255a4633e383b76b9b7e79b97edR203-R207), and updated the logic so only `srcSetBreakpoints` will be used when they are passed in.

File: packages/gatsby-plugin-sharp/src/__tests__/utils.js (hunk 5 above)
Comment (gillkyle) [reply packages/gatsby-plugin-sharp/src/__tests__/utils.js]: Updated this, but a second set of eyes on the logic would be much appreciated 😄

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 4 above)
Comment (ascorbic) [reply packages/gatsby-plugin-sharp/src/utils.js:217]: Omg. TIL!

**Review by ascorbic — APPROVED** [review]

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 7]
```diff
@@ -88,8 +88,27 @@ export function rgbToHex(red, green, blue) {
     .slice(1)}`
 }
 
+const checkIgnoredParameters = (layout, parameters, filepath) => {
+  const ignoredParams = Object.entries(parameters).filter(([_, value]) =>
+    Boolean(value)
+  )
+  if (ignoredParams.length) {
+    console.warn(
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/utils.js:96]: Please use the reporter
```suggestion
    console.warn(
```

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 8]
```diff
@@ -88,8 +88,27 @@ export function rgbToHex(red, green, blue) {
...
+        .map(param => param.join(`: `))
+        .join(
+          `, `
+        )} for the image at ${filepath} are ignored in ${layout} image layouts.`
+    )
+  }
+  return
+}
+
+const DEFAULT_PIXEL_DENSITIES = [0.25, 0.5, 1, 2, 3]
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/utils.js:107]: We should not enable density 3 by default, for constrained images it hardly make sense unless you have a gallery where you have thumbnails and a zoom-in picture. I would add a "highRes" prop to enable this behavior instead.

A good article about this https://blog.twitter.com/engineering/en_us/topics/infrastructure/2019/capping-image-fidelity-on-ultra-high-resolution-devices.html

```suggestion
const DEFAULT_PIXEL_DENSITIES = [0.25, 0.5, 1, 2]
```

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 9]
```diff
@@ -88,8 +88,27 @@ export function rgbToHex(red, green, blue) {
...
+const DEFAULT_FLUID_SIZE = 800
 export const calculateImageSizes = ({
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/utils.js:109]: I would split up this function where fixed & fluid is calculated so it's easier to read

```
if (layout  === 'fixed') {
  sizes = calculateFixedSizes(...)
} elseif (layout === 'fluid' || layout === 'constrained') {
  sizes = calculatedFluidSizes()
}

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 10]
```diff
@@ -98,77 +117,108 @@ export const calculateImageSizes = ({
...
+  // Sort, dedupe and ensure there's a 1
+  const densities = Array.from(
+    new Set([1, ...(outputPixelDensities || DEFAULT_PIXEL_DENSITIES)])
+  ).sort()
+
   if (layout === `fixed`) {
-    // FIXED
-    // if no width is passed, we need to resize the image based on the passed height
-    const fixedDimension = width === undefined ? `height` : `width`
+    checkIgnoredParameters(`fixed`, { maxWidth, maxHeight }, file.absolutePath)
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/utils.js:141]: I would call this function differently, it's doing more than checking
```suggestion
    warnIgnoredParametersUsage(`fixed`, { maxWidth, maxHeight }, file.absolutePath)
```

File: packages/gatsby-plugin-sharp/src/utils.js [hunk 11]
```diff
@@ -98,77 +117,108 @@ export const calculateImageSizes = ({
...
+
+    // if it still hasn't been found, calculate maxWidth from the derived maxHeight
+    if (!maxWidth) {
+      maxWidth = maxHeight * aspectRatio
     }
 
     // Create sizes (in width) for the image if no custom breakpoints are
     // provided. If the max width of the container for the rendered markdown file
-    // is 800px, the sizes would then be: 200, 400, 800, 1200, 1600.
+    // is 800px, the sizes would then be: 200, 400, 800, 1600, 2400 (based off of
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/utils.js:194]: When we drop 3x
```suggestion
    // is 800px, the sizes would then be: 200, 400, 800, 1600 (based off of
```

File: packages/gatsby-plugin-sharp/src/utils.js (hunk 8 above)
Comment (ascorbic) [reply packages/gatsby-plugin-sharp/src/utils.js:107]: Rather than a highRes prop, we let the user pass in an array of densities. Agreed that the default should be a max of 2 though

Comment (wardpeet) [reply packages/gatsby-plugin-sharp/src/utils.js:107]: highRes is food for thought cause it might be easier for people to understand that than srcBreakPoints. Again, not for now ;)

**Comment by ascorbic** [comment]
I've done Ward's suggestions so we can get this in.
